File: app/android/app/src/main/java/com/harvestschools/app/HarvestBrowserPlugin.java
package com.harvestschools.app;

import android.webkit.CookieManager;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

@CapacitorPlugin(name = "HarvestBrowser")
public class HarvestBrowserPlugin extends Plugin implements HarvestBrowserDialog.Listener {

    private HarvestBrowserDialog dialog;
    private boolean isPresented = false;

    private HarvestBrowserDialog.Chrome readChrome(PluginCall call) {
        HarvestBrowserDialog.Chrome chrome = new HarvestBrowserDialog.Chrome();

        chrome.showUrlBar = Boolean.TRUE.equals(call.getBoolean("showUrlBar", true));
        chrome.collapseUrlBarOnScroll = Boolean.TRUE.equals(call.getBoolean("collapseUrlBarOnScroll", true));
        chrome.showBack = Boolean.TRUE.equals(call.getBoolean("showBack", true));
        chrome.showForward = Boolean.TRUE.equals(call.getBoolean("showForward", false));
        chrome.showReload = Boolean.TRUE.equals(call.getBoolean("showReload", true));
        chrome.showShare = Boolean.TRUE.equals(call.getBoolean("showShare", false));
        chrome.showClose = Boolean.TRUE.equals(call.getBoolean("showClose", true));
        chrome.keepTopInset = Boolean.TRUE.equals(call.getBoolean("keepTopInset", true));

        return chrome;
    }

    private Map<String, String> readHeaders(PluginCall call) {
        Map<String, String> headers = new HashMap<>();
        JSObject provided = call.getObject("headers");

        if (provided != null) {
            Iterator<String> keys = provided.keys();

            while (keys.hasNext()) {
                String key = keys.next();
                String value = provided.getString(key);

                if (value != null) { headers.put(key, value); }
            }
        }

        return headers;
    }

    @PluginMethod
    public void open(PluginCall call) {
        String url = call.getString("url");

        if (url == null || url.isEmpty()) {
            call.reject("A url is required");

            return;
        }

        Map<String, String> headers = readHeaders(call);
        HarvestBrowserDialog.Chrome chrome = readChrome(call);
        boolean startHidden = Boolean.TRUE.equals(call.getBoolean("hidden", false));

        getActivity().runOnUiThread(() -> {
            dismissIfNeeded();

            dialog = new HarvestBrowserDialog(getActivity(), url, headers, chrome, this);

            if (startHidden) {
                dialog.create();
                isPresented = false;
            } else {
                dialog.show();
                isPresented = true;
            }

            call.resolve();
        });
    }

    private void dismissIfNeeded() {
        if (dialog != null) {
            try {
                dialog.dismiss();
            } catch (Exception ignored) {
                // Already gone.
            }
        }

        dialog = null;
        isPresented = false;
    }

    @PluginMethod
    public void close(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            dismissIfNeeded();

            notifyListeners("browserClosed", new JSObject());

            call.resolve();
        });
    }

    @PluginMethod
    public void show(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (dialog != null && !isPresented) {
                dialog.show();
                isPresented = true;
            }

            call.resolve();
        });
    }

    @PluginMethod
    public void hide(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (dialog != null && isPresented) {
                dialog.hide();
                isPresented = false;
            }

            call.resolve();
        });
    }

    @PluginMethod
    public void executeScript(PluginCall call) {
        String code = call.getString("code");

        if (code == null) {
            call.reject("code is required");

            return;
        }

        getActivity().runOnUiThread(() -> {
            if (dialog == null || dialog.getWebView() == null) {
                call.reject("No browser is open");

                return;
            }

            dialog.getWebView().evaluateJavascript(code, value -> call.resolve());
        });
    }

    @PluginMethod
    public void getCookies(PluginCall call) {
        String url = call.getString("url");

        JSObject result = new JSObject();

        if (url == null) {
            call.resolve(result);

            return;
        }

        String raw = CookieManager.getInstance().getCookie(url);

        if (raw != null) {
            for (String pair : raw.split(";")) {
                int separator = pair.indexOf('=');

                if (separator > 0) {
                    result.put(pair.substring(0, separator).trim(), pair.substring(separator + 1).trim());
                }
            }
        }

        call.resolve(result);
    }

    @PluginMethod
    public void clearCookies(PluginCall call) {
        String url = call.getString("url");
        CookieManager manager = CookieManager.getInstance();

        if (url == null) {
            manager.removeAllCookies(value -> call.resolve());

            return;
        }

        String raw = manager.getCookie(url);

        if (raw != null) {
            for (String pair : raw.split(";")) {
                int separator = pair.indexOf('=');

                if (separator > 0) {
                    String name = pair.substring(0, separator).trim();

                    manager.setCookie(url, name + "=; Max-Age=0; Path=/");
                }
            }
        }

        manager.flush();

        call.resolve();
    }

    @Override
    public void onBrowserClosed() {
        isPresented = false;
        dialog = null;

        notifyListeners("browserClosed", new JSObject());
    }

    @Override
    public void onPageLoaded(String url) {
        JSObject payload = new JSObject();
        payload.put("url", url);

        notifyListeners("browserPageLoaded", payload);
    }

    @Override
    public void onUrlChanged(String url) {
        JSObject payload = new JSObject();
        payload.put("url", url);

        notifyListeners("urlChange", payload);
    }

    @Override
    public void onMessage(String rawJson) {
        JSObject payload = new JSObject();

        try {
            payload.put("detail", JSObject.fromJSONObject(new JSONObject(rawJson)));
        } catch (Exception parseError) {
            payload.put("detail", new JSObject());
            payload.put("rawMessage", rawJson);
        }

        notifyListeners("messageFromWebview", payload);
    }
}

File: app/ios/App/App/HarvestBrowserPlugin.swift
import Foundation
import Capacitor
import UIKit
import WebKit

@objc(HarvestBrowserPlugin)
public class HarvestBrowserPlugin: CAPPlugin, CAPBridgedPlugin, HarvestBrowserControllerDelegate {

    public let identifier = "HarvestBrowserPlugin"

    public let jsName = "HarvestBrowser"

    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executeScript", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCookies", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearCookies", returnType: CAPPluginReturnPromise)
    ]

    private var controller: HarvestBrowserController?
    private var isPresented = false
    private var isVisible = false
    private var hiddenContainer: UIView?

    private func activeWindow() -> UIWindow? {
        return UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive || $0.activationState == .foregroundInactive }?
            .windows
            .first { $0.isKeyWindow }
    }

    private func attachWebViewToWindow(_ webView: WKWebView) -> Bool {
        guard let window = activeWindow() else { return false }

        let frame = CGRect(
            x: window.bounds.maxX + 2048,
            y: 0,
            width: max(window.bounds.width, 1),
            height: max(window.bounds.height, 1)
        )

        let container = hiddenContainer ?? UIView(frame: frame)

        container.frame = frame
        container.backgroundColor = .clear
        container.isOpaque = false
        container.isUserInteractionEnabled = false
        container.clipsToBounds = true
        container.accessibilityElementsHidden = true

        if container.superview !== window {
            container.removeFromSuperview()
            window.addSubview(container)
        }

        hiddenContainer = container

        webView.removeFromSuperview()
        webView.translatesAutoresizingMaskIntoConstraints = true
        webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        webView.frame = container.bounds
        webView.isUserInteractionEnabled = false

        container.addSubview(webView)

        return true
    }

    private func releaseHiddenContainer() {
        hiddenContainer?.removeFromSuperview()
        hiddenContainer = nil
    }

    private func readChrome(from call: CAPPluginCall) -> HarvestBrowserChrome {
        var chrome = HarvestBrowserChrome()

        chrome.showUrlBar = call.getBool("showUrlBar", true)
        chrome.collapseUrlBarOnScroll = call.getBool("collapseUrlBarOnScroll", true)
        chrome.showBack = call.getBool("showBack", true)
        chrome.showForward = call.getBool("showForward", false)
        chrome.showReload = call.getBool("showReload", true)
        chrome.showShare = call.getBool("showShare", false)
        chrome.showClose = call.getBool("showClose", true)
        chrome.keepTopInset = call.getBool("keepTopInset", true)

        return chrome
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let address = call.getString("url"), let url = URL(string: address) else {
            call.reject("A url is required")

            return
        }

        var headers: [String: String] = [:]

        call.getObject("headers")?.forEach { key, value in
            if let text = value as? String { headers[key] = text }
        }
        let startHidden = call.getBool("hidden", false)
        let chrome = readChrome(from: call)

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.dismissIfNeeded {
                let created = HarvestBrowserController(url: url, headers: headers, chrome: chrome)
                created.delegate = self

                self.controller = created

                created.loadViewIfNeeded()

                if startHidden {
                    self.isPresented = false
                    self.isVisible = false

                    if self.attachWebViewToWindow(created.webView) {
                        call.resolve()
                    } else {
                        call.reject("No active window to load in")
                    }

                    return
                }

                self.presentController(created) { call.resolve() }
            }
        }
    }

    private func presentController(_ created: HarvestBrowserController, completion: @escaping () -> Void) {
        guard let presenter = bridge?.viewController else {
            completion()

            return
        }

        created.loadViewIfNeeded()

        releaseHiddenContainer()
        created.reattachWebView()

        isPresented = true
        isVisible = true

        created.modalPresentationStyle = .fullScreen
        created.view.alpha = 1
        created.view.isUserInteractionEnabled = true

        presenter.present(created, animated: true) { completion() }
    }

    private func dismissIfNeeded(completion: (() -> Void)? = nil) {
        let wasPresented = isPresented
        let existing = controller

        isPresented = false
        isVisible = false
        controller = nil

        releaseHiddenContainer()

        guard wasPresented, let existing = existing else {
            completion?()

            return
        }

        existing.dismiss(animated: false) { completion?() }
    }

    @objc func close(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.dismissIfNeeded {
                self?.notifyListeners("browserClosed", data: [:])

                call.resolve()
            }
        }
    }

    @objc func show(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let created = self.controller else {
                call.resolve()

                return
            }

            if self.isPresented {
                self.isVisible = true

                call.resolve()

                return
            }

            self.presentController(created) { call.resolve() }
        }
    }

    @objc func hide(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let created = self.controller else {
                call.resolve()

                return
            }

            self.isVisible = false

            let park = {
                _ = self.attachWebViewToWindow(created.webView)

                call.resolve()
            }

            if self.isPresented {
                self.isPresented = false

                created.dismiss(animated: false) { park() }
            } else {
                park()
            }
        }
    }

    @objc func executeScript(_ call: CAPPluginCall) {
        guard let code = call.getString("code") else {
            call.reject("code is required")

            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let webView = self?.controller?.webView else {
                call.reject("No browser is open")

                return
            }

            webView.evaluateJavaScript(code) { _, error in
                if let error = error {
                    call.reject(error.localizedDescription)
                } else {
                    call.resolve()
                }
            }
        }
    }

    @objc func getCookies(_ call: CAPPluginCall) {
        guard let address = call.getString("url"), let url = URL(string: address) else {
            call.reject("A url is required")

            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let webView = self?.controller?.webView else {
                call.resolve([:])

                return
            }

            webView.configuration.websiteDataStore.httpCookieStore.getAllCookies { cookies in
                var jar: [String: String] = [:]

                cookies.forEach { cookie in
                    if let host = url.host, host.hasSuffix(cookie.domain) || cookie.domain.hasSuffix(host) {
                        jar[cookie.name] = cookie.value
                    }
                }

                call.resolve(jar)
            }
        }
    }

    @objc func clearCookies(_ call: CAPPluginCall) {
        let address = call.getString("url")

        DispatchQueue.main.async { [weak self] in
            guard let webView = self?.controller?.webView else {
                call.resolve()

                return
            }

            let store = webView.configuration.websiteDataStore.httpCookieStore

            store.getAllCookies { cookies in
                let host = address.flatMap { URL(string: $0)?.host }

                let group = DispatchGroup()

                cookies.forEach { cookie in
                    let matches = host == nil || host!.hasSuffix(cookie.domain) || cookie.domain.hasSuffix(host!)

                    if matches {
                        group.enter()

                        store.delete(cookie) { group.leave() }
                    }
                }

                group.notify(queue: .main) { call.resolve() }
            }
        }
    }

    func harvestBrowserDidClose() {
        isPresented = false
        isVisible = false

        releaseHiddenContainer()

        controller = nil

        notifyListeners("browserClosed", data: [:])
    }

    func harvestBrowserDidLoadPage(url: String) {
        notifyListeners("browserPageLoaded", data: ["url": url])
    }

    func harvestBrowserDidChangeUrl(url: String) {
        notifyListeners("urlChange", data: ["url": url])
    }

    func harvestBrowserDidReceiveMessage(_ detail: [String: Any]) {
        notifyListeners("messageFromWebview", data: ["detail": detail])
    }
}
